### ISSUE TITLE
EEPROMs not being changed when accessed by Transposer

### ISSUE THREAD

**Issue by bibo38** [MightyPirates/OpenComputers#3253]
**TLDR** Changing an EEPROM (e.g. label or data) and pulling it out of a computer using a transposer doesn't keep the EEPROM changes.

## Reproduction

Set up a Transposer which is connected to a computer and a chest:
![Bug Setup](https://user-images.githubusercontent.com/1812304/77680325-18caac80-6f94-11ea-9f37-2281014ba2eb.png)

Then install OpenOS on the computer, use the default Lua BIOS and create a `test.lua` file with the following contents:
```lua
component = require("component")

component.eeprom.setLabel("Testname")
component.eeprom.set("Will not work")
print(component.eeprom.getLabel())
component.transposer.transferItem(
  5,  -- Side connected to the computer (may need to be changed)
  1,  -- Chest
  1,  -- 1 EEPROM
  10, -- EEPROM Slot in the computer
  1   -- Chest slot
)
```

Run `test` in the console

### Expected behavior

The chest contains an EEPROM with the name `Testname`, which cannot be used to boot a PC.

### Actual behavior

The chest contains the default Lua BIOS from the Computer. It has the default label and works without a problem.

## Notes

- I didn't tested this with harddrives or floppy disks
- Also didn't tested it with other mods pulling items out of the computer
- If you remove the transposer code, then the tooltip of the EEPROM in the computer shows the wrong Label, but when you remove it (manually), it is changed to the correct label.

**Comment by bibo38** [comment]
Are you already working on this @payonel ? I would also be interested to fix this issue and send a PR. If you have hints where the bug might be located, i would be interested in those.

**Comment by HateMeLikeAPro** [comment]
This seems to not only be a problem with EEPROM, I have a "deploy station" that basically deploys an image to a HDD inserted into a dispenser. This is routed using a transposer. And when I `shell.execute("label set " .. addr .. " SomeLabel")` it only works if I pull out the HDD myself (opening the computer case and pulling the hdd from the inventory into my player inventory). If the transposer does it, the hdd seems to have been wrote to (the fs got formatted and a default install of openOS  is on there) but the label isn't there anymore

**Comment by bibo38** [comment]
> the hdd seems to have been wrote to (the fs got formatted and a default install of openOS is on there) but the label isn't there anymore

Probably, because the fs changes are not metadata, but saved directly as files to disk. The main Problem with the metadata is, that it is only written to the item, when it is extracted. The problem lies in the fact, that in Minecraft you can simulate an insertion/extraction. This is helpful, because not all inventories act like chests and you want to avoid pulling to few or too many items to prevent item losses or duplication. OpenComputers only saves the metadata to the item, when it is not a simulated extraction. This behaviour is fine, but the transfer method of the transposer breaks this behaviour by using the items from the simulated extraction and after the insertion is done doing the actual extraction and discarding the extracted items (which contain the correct metadata).

I haven't tried it, but I think, that the problem can be fixed by using another mod to pull items from the computer. Then use computers to control the mod. Especially having a little AE2 Network could work by controlling the importer/exporter.

**Comment by payonel** [comment]
this is on my short list of bugs to fix, there will be on update on this ticket. thanks @bibo38 for your perspective on the matter

**Comment by JujuLynx** [comment]
So I've been doing some testing on this and found something very interesting!

First point: Everything "technically" works, but there is still a bug. You have to leave the EEPROM in the case for a specific amount of time before removing it. You can observe this yourself. Flash an EEPROM and change the label to something like 'Test'. You will see the changes won't take effect immediately, but give it time and it will change eventually.

How long do you have to wait? Well, that's the interesting part - you have to wait for the autosave.

I also tested this out and like clockwork, the EEPROM will change at the 45 second mark regardless of when you initiate the changes. So depending on when you do it, you could be waiting 5 seconds, or up to 45 seconds for it to work. Then you can safely take the EEPROM out of the case with a Transposer.

As a workaround, include `os.sleep(45)` after making any changes. It's not desirable, but the only fix until this is patched.

As a side note, if you check the EEPROM with a computer directly, such as `component.eeprom.getLabel()` the computer itself will see the new label, but hovering your mouse over the EEPROM will show the old label until the autosave comes around.

Another interesting note is that if you place the EEPROM manually you see the same behaviour, but if you remove the EEPROM yourself you will see the changes immediately without waiting for the autosave. I don't have intimate knowledge of the inner workings of OpenComputers, but I suspect it's because the Transposer is accessing the EEPROM from a different location (in memory, perhaps?) than where a player removing it would. I believe this is where the root cause of the bug is.

I haven't done extensive testing, but I've seen the same behaviour with the Floppy Disk when it comes to labels. I suspect that this behaviour in fact affects ALL items, but it's how the Transposer interacts with items is the culprit (or how OC buffers and flushes data).

Update: I tried replicating this issue with a Floppy Disk and it seems to have the same issues. The only difference is I have `bufferChanges = false`, the computer will write a file to it immediately and it will work even if you remove it with the Transposer. I suspect that with `bufferChanges = true`, changes to the disk will be affected in a similar way. But the issues with changing the label are certainly identical with Floppy Disks.

**Comment by bibo38** [comment]
> but if you remove the EEPROM yourself you will see the changes immediately without waiting for the autosave. I don't have intimate knowledge of the inner workings of OpenComputers, but I suspect it's because the Transposer is accessing the EEPROM from a different location (in memory, perhaps?) than where a player removing it would.

Take a look at #3255. The problem is that the Transposer uses the Itemstack of an simulated extraction and throws the real item stack of the non simulated extraction away. But OpenComputers saves metadata like labels to the NBT of the item only on non simulated extractions (and on the autosave).

So it should be possible to use some other mod to extract the item, that does use the non simulated Itemstack (that has the correct NBT). And then controlling this extraction via some redstone or smth. else from OpenComputers.

> I suspect that with bufferChanges = true, changes to the disk will be affected in a similar way.

Probably not. This issue is about the NBT not being updated, but as harddisks (and i think also floppies) store their data on the filesystem. So this issue shouldn't occur, as we don't change any NBT related data when the data on the floppy/harddisk is changed.

**Comment by FreezePhoenix** [comment]
Turning off the computer and waiting 2 seconds works, instead of waiting 45 seconds.
